New concept First-Class Functions

antklim opened the pull request:
Hi exercism team.

I'm not sure whether this concept is already planned or being worked on. And I apologise if I'm creating inconveniences. 

I was learning different languages by reading books and using your platform. I found that the first-class functions are a powerful tool that is used by developers every day, but sometimes we don't even realise it. I would like to share my knowledge and expertise on this subject. That is a reason to create this PR.

I'm sorry for opening such a big pull request. I wasn't sure what will be the best way to create a new concept and support it with the practice exercise. Please let me know if you would like it to be split into smaller PRs. 

Thanks,
Anton

andrerfcsantos commented on exercises/concept/expenses/.meta/exemplar.go:
```diff
@@ -0,0 +1,127 @@
+func (x byCategoryName) Less(i, j int) bool { return x[i].Category < x[j].Category }
+func (x byCategoryName) Swap(i, j int)      { x[i], x[j] = x[j], x[i] }
+
+func byDatePeriod(p DatePeriod) func(Record) bool {
```
I really like the idea of `byDatePeriod` and `byCategory` as good exercises to practice returning functions.

However, I think they should be on their own task. In another comment, I suggested the making of `Filter` also be a task and be the first one, and I think these functions could be the tasks next to that one.

That way, we have these tasks acting as "checkpoints" before the task where we ask the student to write `Total`.

andrerfcsantos commented on exercises/concept/expenses/.docs/instructions.md:
````diff
@@ -0,0 +1,115 @@
+// Output: 0
+```
+
+## 2. Find the top N expenses categories in period
````
I think this task should be removed. Adds considerable amount of busy work like sorting, that doesn't necessarily help the student learn more about functions as first-class values.

See the other comments for suggestions on additional tasks. I believe with those additional tasks, this one becomes unnecessary.

andrerfcsantos reviewed (changes requested):
Thanks for this!

Since this is indeed a big PR, took us a while to get back to you, sorry for that.

First class functions is something that we had planned on our [Roadmap](https://exercism.org/docs/tracks/go/roadmap#h-potential-future-concepts) as a potential future concept and not something that was meant to me implemented soon. Concepts that we mean to implement all have their own issues that are being tracked [here](https://github.com/exercism/go/issues/1886). The advantage of having an issue associated to each concept exercise is that we can discuss some things on the issue itself and agree on a general implementation before actually implementing it.

Despite this PR not following the expected flow for a concept exercise, I think we can accommodate this in our current concept tree!

Overall the PR is really good. It shows that you took your time with it, so thank you for that!

I did some initial review requesting changes, let me know what you think of them.

antklim commented on concepts/first-class-functions/about.md:
```diff
@@ -0,0 +1,89 @@
+
+Go supports functions as first-class values. It means a function can be an argument to other functions, returned by another function, and assigned to variables.
+
+It is possible thanks to the function types in Go. A function type denotes the set of all functions with the same sequence of parameter types and the same sequence of result types. User-defined types can be declared on top of function types.
```
Ok, got it. I will update examples. Thanks for the help.

antklim commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,49 @@
+
+// Filter creates a new records collection by applying the predicate function to
+// records and keeping the records for which the function returns true.
+func (rr Records) Filter(f func(Record) bool) Records {
```
Not a problem, I will update the exercise accordingly.

antklim commented on exercises/concept/expenses/.docs/instructions.md:
````diff
@@ -0,0 +1,115 @@
+Implement the `Total` function to return a sum of expenses in period:
+
+```go
+oct2000 := DatePeriod{From: "2000-10-01", To: "2000-10-31"}
````
Sure, I'll alter the examples to show the creation of `time.Time`.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+
+// Filter creates a new records collection by applying the predicate function to
+// records and keeping the records for which the function returns true.
+func Filter(in []Record, f func(Record) bool) (out []Record) {
```
```suggestion
func Filter(in []Record, f func(Record) bool) []Record {
```

As much as possible, we try to discourage the use of named returns.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+// Filter creates a new records collection by applying the predicate function to
+// records and keeping the records for which the function returns true.
+func Filter(in []Record, f func(Record) bool) (out []Record) {
+	panic("Filter not implemented")
```
```suggestion
	panic("Please implement the Filter function.")
```

Just to make the message look consistent with others from other exercises.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+// ByDatePeriod returns predicate function. The predicate returns true when
+// Record.Date is within the date period.
+func ByDatePeriod(p DatePeriod) func(Record) bool {
+	panic("ByDatePeriod not implemented")
```
```suggestion
	panic("Please implement the ByDatePeriod function.")
```

Just to make the message look consistent with others from other exercises.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+// ByCategory returns predicate function. The predicate returns true when
+// Record.Category is equal to the provided category.
+func ByCategory(c string) func(Record) bool {
+	panic("ByCategory not implemented")
```
```suggestion
	panic("Please implement the ByCategory function.")
```

Just to make the message look consistent with others from other exercises.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+// Total returns total amount of expenses in collection of records rr
+// within time period p.
+func Total(in []Record, p DatePeriod) float64 {
+	panic("Total not implemented")
```
```suggestion
	panic("Please implement the Total function.")
```

Just to make the message look consistent with others from other exercises.

andrerfcsantos commented on exercises/concept/expenses/expenses.go:
```diff
@@ -0,0 +1,46 @@
+// CategoryExpenses returns total amount of expenses in category c. It returns
+// error when a category is not present in expenses records.
+func CategoryExpenses(in []Record, p DatePeriod, c string) (float64, error) {
+	panic("CategoryExpenses not implemented")
```
```suggestion
	panic("Please implement the CategoryExpenses function.")
```

Just to make the message look consistent with others from other exercises.

andrerfcsantos commented on concepts/first-class-functions/about.md:
````diff
@@ -0,0 +1,76 @@
+}
+```
+
+Another powerful tool that is available thanks to first-class functions support is anonymous functions. Anonymous functions are defined at treir point of use, without a name following the `func` keyword. Such functions have access to the variables of the enclosing function. For example:
````
```suggestion
Another powerful tool that is available thanks to first-class functions support is anonymous functions. Anonymous functions are defined at their point of use, without a name following the `func` keyword. Such functions have access to the variables of the enclosing function. For example:
```

andrerfcsantos commented on exercises/concept/expenses/.meta/design.md:
```diff
@@ -0,0 +1,23 @@
+
+## Out of scope
+
+## Concepts
+
```
```suggestion
## Concepts

- `first-class-functions`

```

andrerfcsantos reviewed (changes requested):
This is looking ever better!

The main suggestion here is to replace the use of real times to a simpler version of time that can be represented by just a number (i.e, number of days). I also suggested other minor fixes. After these changes, I think we can launch the exercise.

antklim commented:
Hi @andrerfcsantos, thanks a lot for the reviews! I hope I addressed all the comments now. Please let me know if there are more improvements that need to be done. 
Also, I have a question. What status should the `expenses` exercise have? I put it to `wip` in `config.json`. Should it be changed to `beta` or any other status?
Thanks

andrerfcsantos reviewed (approved):
Since this PR is going for a while and to not take much of your time and effort, I just did some final touches to the exercise. The main thing was to update the hints to match the changes made to the exercise and add some clarifications in the text.

Thank you for all the work and patience in getting this exercise ready!
